File: packages/core-python/jsonmd/table.py
"""Markdown table rendering utilities."""

from __future__ import annotations

from typing import Any


def _escape_cell(value: Any) -> str:
    """Escape a value for use in a Markdown table cell."""
    if value is None:
        return ""
    s = str(value)
    return s.replace("|", "\\|").replace("\n", " ")


def _format_value(value: Any) -> str:
    """Format a value for display in a Markdown cell."""
    if value is None:
        return ""
    if isinstance(value, bool):
        return str(value).lower()
    if isinstance(value, (int, float)):
        return str(value)
    if isinstance(value, str):
        return value
    import json
    return json.dumps(value)


def is_primitive(value: Any) -> bool:
    """Check if a value is a primitive (str, int, float, bool, None)."""
    return value is None or isinstance(value, (str, int, float, bool))


def is_flat(obj: dict) -> bool:
    """Check if a dict has only primitive values."""
    return all(is_primitive(v) for v in obj.values())


def to_table(rows: list[dict]) -> str:
    """Convert a list of flat dicts to a Markdown table."""
    if not rows:
        return ""

    # Collect union of keys in first-occurrence order
    seen: dict[str, None] = {}
    for row in rows:
        for key in row:
            if key not in seen:
                seen[key] = None
    keys = list(seen.keys())

    lines = []
    lines.append("| " + " | ".join(keys) + " |")
    lines.append("| " + " | ".join("---" for _ in keys) + " |")

    for row in rows:
        cells = [_escape_cell(_format_value(row.get(k))) for k in keys]
        lines.append("| " + " | ".join(cells) + " |")

    return "\n".join(lines)


def to_key_value_table(obj: dict) -> str:
    """Convert a flat dict to a 2-column Key | Value table."""
    if not obj:
        return ""

    lines = []
    lines.append("| Key | Value |")
    lines.append("| --- | --- |")

    for key, value in obj.items():
        lines.append(f"| {_escape_cell(key)} | {_escape_cell(_format_value(value))} |")

    return "\n".join(lines)


def to_bullet_list(items: list) -> str:
    """Convert a list of primitives to a Markdown bullet list."""
    return "\n".join(f"- {_format_value(item)}" for item in items)

File: packages/core-js/src/table.ts
import type { JsonPrimitive, JsonValue } from './types.js';

function escapeCell(value: unknown): string {
  if (value === null || value === undefined) return '';
  const str = String(value);
  return str.replace(/\|/g, '\\|').replace(/\n/g, ' ');
}

function formatValue(value: unknown): string {
  if (value === null || value === undefined) return '';
  if (typeof value === 'boolean') return String(value);
  if (typeof value === 'number') return String(value);
  if (typeof value === 'string') return value;
  return JSON.stringify(value);
}

/**
 * Convert an array of flat objects to a Markdown table.
 */
export function toTable(rows: Record<string, unknown>[]): string {
  if (rows.length === 0) return '';

  // Collect union of keys in first-occurrence order
  const keySet = new Set<string>();
  for (const row of rows) {
    for (const key of Object.keys(row)) {
      keySet.add(key);
    }
  }
  const keys = Array.from(keySet);

  const lines: string[] = [];

  // Header
  lines.push('| ' + keys.join(' | ') + ' |');
  // Separator
  lines.push('| ' + keys.map(() => '---').join(' | ') + ' |');
  // Data rows
  for (const row of rows) {
    const cells = keys.map((k) => escapeCell(formatValue(row[k])));
    lines.push('| ' + cells.join(' | ') + ' |');
  }

  return lines.join('\n');
}

/**
 * Convert a flat object to a 2-column Key | Value table.
 */
export function toKeyValueTable(obj: Record<string, unknown>): string {
  const entries = Object.entries(obj);
  if (entries.length === 0) return '';

  const lines: string[] = [];
  lines.push('| Key | Value |');
  lines.push('| --- | --- |');

  for (const [key, value] of entries) {
    lines.push(`| ${escapeCell(key)} | ${escapeCell(formatValue(value))} |`);
  }

  return lines.join('\n');
}

/**
 * Convert an array of primitives to a bullet list.
 */
export function toBulletList(items: JsonPrimitive[]): string {
  return items.map((item) => `- ${formatValue(item)}`).join('\n');
}

/**
 * Check if a value is a primitive (string, number, boolean, null).
 */
export function isPrimitive(value: unknown): value is JsonPrimitive {
  return value === null || typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean';
}

/**
 * Check if an object has only primitive values.
 */
export function isFlat(obj: Record<string, unknown>): boolean {
  return Object.values(obj).every(isPrimitive);
}
